### DISCUSSION TITLE
Question: resize without maintaining image aspect ratio

### DISCUSSION THREAD

**Discussion by mxmaxime** [libvips/libvips#2588] [Q&A]
Hello,
First of all, thanks for the lib!
Sorry to open an issue for that, but I have a question about the resize function.
[From the doc](https://www.libvips.org/API/current/libvips-resample.html#vips-resize): "vips_resize() normally maintains the image aspect ratio.", but I'd like to force the image to have a fixed size. I'm using scale and vscale but the end result is not what I want because it seems to respect the image ratio...

C++ code:

```cpp
// I want the image to be 300x200
double hscale = 200.0 / (double) in.height();
double vscale = 300.0 / (double) in.width();
std::cout << "h scale= " << hscale << " vscale= " << vscale << std::endl;

VImage out = in.resize(hscale,
        VImage::option()->set("vscale", vscale));
```
I know the thumbnail function has a parameter to force the image to fit the target by breaking the image ratio, but what about the resize function?
I tried to look at the code, but I can't figure it out (I'm a newbie in the C/C++ world).

Thanks a lot!

**Comment by jcupitt** [comment]
Hi @mxmaxime, `resize` should let you break the aspect ratio. I see:

```
$ vips resize k2.jpg x.jpg 0.1 --vscale 0.001
$ vipsheader k2.jpg x.jpg
k2.jpg: 1450x2048 uchar, 3 bands, srgb, jpegload
x.jpg: 145x2 uchar, 3 bands, srgb, jpegload
```

Your code looks OK at a quick glance. If you post a complete example that shows the problem I can try running it.

**Comment by jcupitt** [comment]
... I expect you know, but `resize` is quite low-level -- on its own it won't give good results for images containing transparent elements, for example.

Unless you have some very specific need, it's best to use `thumbnail`, if you can.

**Comment by mxmaxime** [comment]
Hello,
Thanks for your quick reply, it unlocked me. Ok, I didn't know it's low-level, I use thumbnail functions to resize my image now.
As I saw, it uses the resize method under the hood and this function does anti aliasing processing which is a requirement so I guess I can use thumbnail.
> vips_resize() normally uses VIPS_KERNEL_LANCZOS3 for the final reduce, you can change this with kernel.
> Make a thumbnail from a file. Shrinking is done in three stages: using any shrink-on-load features available in the file import library, using a block shrink, and using a lanczos3 shrink. At least the final 200% is done with lanczos3. The output should be high quality, and the operation should be quick.

However I have a question: **how can I create a thumbnail from a buffer?**
Here's the code that I have (I try to simplify things).

```cpp
// from mqtt stuff
auto payload = msg->get_payload();
// I know how to load the image from the buffer (save works as expected, the image is OK)
auto in = VImage::new_from_buffer(payload.data(), payload.size(), nullptr);
// but.. what about thumbnail? I'd need something like:
auto out = VImage::thumbnail_buffer(payload.data(), payload.size(), 300,
    VImage::option()->set("width", 300)->set("size", VIPS_SIZE_FORCE)->set("no_rotate", true));

// as of today I use:
in.thumbnail_image(300, VImage::option()->set("width", 300)->set("size", VIPS_SIZE_FORCE)->set("no_rotate", true));
```
I saw the method [thumbnail_buffer](https://www.libvips.org/API/8.11/cpp/classVImage.html#afa428a16a001d1e1032a62daa4881c85) but it asks a `VipsBlob*`.
Also, I'm wondering if there's a "good" way in term of performance to do so.

Huge thanks!

**Reply by jcupitt** [reply]
You can make a VipsBlob with this:

https://www.libvips.org/API/current/libvips-type.html#vips-blob-new

Example of use here:

https://github.com/libvips/libvips/blob/master/libvips/resample/thumbnail.c#L1483-L1501

VipsBlob should probably be wrapped for the C++ API, but you can always use the C one.
